diff --git a/CMake/HPHPFindLibs.cmake b/CMake/HPHPFindLibs.cmake
index b5514ff..337c41b 100644
--- a/CMake/HPHPFindLibs.cmake
+++ b/CMake/HPHPFindLibs.cmake
@@ -60,10 +60,12 @@ endif()
 
 # mysql checks - if we're using async mysql, we use webscalesqlclient from
 # third-party/ instead
+include_directories(
+  ${TP_DIR}/re2/src/
+  ${TP_DIR}/squangle/src/
+)
 if (ENABLE_ASYNC_MYSQL)
   include_directories(
-    ${TP_DIR}/re2/src/
-    ${TP_DIR}/squangle/src/
     ${TP_DIR}/webscalesqlclient/src/include/
   )
   set(MYSQL_CLIENT_LIB_DIR ${TP_DIR}/webscalesqlclient/src/)
